### ISSUE TITLE
Regions Level-3

### ISSUE THREAD

**Issue by GoogleCodeExporter** [opencaching/opencaching-pl#103]
```
There are 55 regions in the nuts_codes table that match code like 'RO%'.
There are 55 regions in the nuts_layers table that match code like 'RO%'.
Joining the tables using code produces 55 results, so the correspondence is 
correct. 

However only 8 regions are displayed in the dropdown box. 
Most important problem: none of the counties are in it.
Pressing zone from coordinates yields empty results. 

See already published caches. They should all match Argeş county, that is 
region RO311. None do so. 
```

Original issue reported on code.google.com by `andrixnet` on 9 Apr 2014 at 7:56

**Comment by kojoty** [comment]
@andrixnet, please review this and if necessary add summarize what should be done here and reopen then. Thanks.

**Comment by andrixnet** [comment]
Most OC sites use NUTS layer 3 data, as this is relevant for their country's context.
However relevant region data for Romania is equivalent to data in NUTS layer 4.

**Comment by andrixnet** [comment]
According to my research: 
Countries are at level 0 (numbering in table)
macro regions are at level 1
provinces are at level 2 (most EU sites, CA, US, MX)
counties are at level 3 (required by RO since romanian "judeţ" is at this level)

Current code recognizes provinces at level 2 only (tested with new data for US).

**Comment by kojoty** [comment]
@andrixnet 
Please answer my question in context of OCUS: 
- OCPL/RO/NL used only two levels of regions: country + some province (NUTS-level 2 or 3)

- for OCUS we have:  country (US) +state + county 

Do OCUS need all 3 levels (country+state+county) or only set of 2 from those three... Or maybe we should merge states and counties and publish something like: US - CA /San Francisco (county)

**Comment by kojoty** [comment]
@andrixnet , my next update to this issue - draft of my idea:
@deg-pl, @harrieklomp  FYI as well

- lets make one new aka-nuts table with structure:
   - global unique code of country / region [text]
   - level (flat structure like: 0=countries | 1=macroregion/state (optional) 2=region)
   - upper-level code (country for states, state for  regions etc.)
   - region name (regions lev.2 should contains name of the macroregion code like: "CA/SanFrancisco")
   - shape (geo-data)

- every cache has assigned its: country + region (+optionally macroregion if present) in cache_location tables  

- I'm able to prepare such data to import into tables - for now I think I have enough experience this - I've found good source of such geo data based on official gov. institutions

When all will be ready (hope after the weekend) we will be able to migrate all nodes to this.

**Comment by andrixnet** [comment]
Since our code only presents the user only 2 informations: country and region, what you describe above doesn't seem to simplify it. 

Existing data already has global unique code for country and region (such as US = USA (country)  and US12=Kansas (state/province). Or PL  = Poland and PL12=[Lubuskie](https://en.wikipedia.org/wiki/Lubusz_Voivodeship). 

_Haven't seen (so far) any interest in macro regions such as "south-west", so that level of data would it really be necessary to have and process?_

_Also, haven't seen any interest at county level (where much more geo data is required). 
(example: [Słubice County](https://en.wikipedia.org/wiki/S%C5%82ubice_County) of Lubusz Voivodeship)_

_The same goes for San Francisco County of California ... (from above example CA/San Francisco)_

The equivalent for RO according to existing data: RO = Romania and RO123 = [Arges](https://en.wikipedia.org/wiki/Arge%C8%99_County). I know the english terminology is "county", but this is the same as polish Voivodeship, german Land and US state.
At RO12 we have "South" for example, which is of zero relevance. 

(region codes are not actual, just examples). 

----------------------------------

Thus I see two ways of handling this: 

1. keep geo data, change PHP code:
- keep geo data and region codes as they are. 
- PHP code handling countries probably does not need changes
- PHP code  handling regions needs to: 
  - make use of all 4 levels of data (if exists) as it is now (though some will be useless)
  - make the level configurable (for each site)

2. process geo data, change PHP code:
- geo data to contain, for each country, only 2 levels: country outline and province/state outline.
- country outline code is already 2 characters long, unique, based on ISO
- province/state outline may be variable in length (depending on country), but at least 3 characters long. (usually 4 or 5)
- level column is no longer needed (just use code length as key)
- adm/code 1-4 columns are no longer needed (and the PHP code handling them)
- caches table can have 2 columns to reference country and province/state

Both scenarios may also handle US ZIP codes using :
1. a dedicated table with shape data
2. integrate shape data into country/region table and use a separate column with ZIP code (country/region data having this column as NULL)

---------------------------------

**Comment by rapotek** [comment]
Long ago I recommended using iso3166 -1 and -2 codes (even made some tests in the subject). They do not include US counties, but may be sufficient for the rest of the world and counties may be imported otherwise.
As for level in proposed table: why not base on OSM data, and use its "admin_level" attribute associated with border shapes? The attribute exists regardless of iso, nuts, fips or ansi values.

**Comment by kojoty** [comment]
@andrixnet 
I think I was not explained it well and we don't understand each other :)

My general opinion is that we need 3 levels: L0=country + L1=state + L2=local-region

For most countries like Pl or RO L1 is not necessary - for PL:  L0=PL + PL12=Lubuskie is enough

But for US where Teksas is probably bigger than Poland and Romania together would be good to still have third level like: L0=US (country) L1=USTX (state)  L2=Harries (county) - this is the only reason why I think about the third level (L1) - without this we can just use two levels: country + region.

For most countries we can skip L1 and have just only L0+L2.

I see two cases when we need these data:
- for describing of the cache (case viewcache) - here we can use country + L2 name of the region
- for grouping of the cache (case search by region) - here we can use country + L1 (if present) or L2 otherwise 

We can chat over IM or call to find the right vision :)

**Comment by andrixnet** [comment]
AFAIK in USA states are very important and ZIP code is more important then county. But let's take it up with them.

**Comment by kojoty** [comment]
> Long ago I recommended using iso3166 -1 and -2 codes (even made some tests in the subject). They do not include US counties, but may be sufficient for the rest of the world and counties may be imported otherwise.
> As for level in proposed table: why not base on OSM data, and use its "admin_level" attribute associated with border shapes? The attribute exists regardless of iso, nuts, fips or ansi values.

@rapotek, please point me to the source of OSM data - maybe it would be a good idea...

ISO code-s is not a problem in my opinion - For countries we used ISO codes now (two letters version). For regions we use the codes internal only so this is not the case - if we use gov. data (I've found a good source of shapefiles for EU, US, CA, MX for now) it much easy to rely on codes from that files.

**Comment by kojoty** [comment]
@andrixnet ZIP-codes is a separate case we can add it later - I think it should be stored in different table etc. 

BTW. in my opinion this is a little strange case of usage: do you really need to search geocaches by post-code? Because of why?

**Comment by andrixnet** [comment]
@kojoty if we include county data or macro region data, the next step will be searching by that criteria (I'm not saying it, someone will definitely ask for it though).

Regarding ZIP codes: 
- AFAIK Groundspeak has ZIP code criteria
- AFAIK americans make a lot of use out of this information (but have to ask them)
See https://en.wikipedia.org/wiki/ZIP_Code#Other_uses

**Comment by rapotek** [comment]
@kojoty just check `https://overpass-turbo.eu/` for live example. Sample queries:

- all countries with boundaries:
`[out:json];(rel["ISO3166-1"][boundary=administrative];);out tags;`
(the result is big)

- Polish iso3166-2 regions:
`[out:json];(rel["ISO3166-2"~"^PL-.*"][boundary=administrative];);out tags;`

- US counties:
`[out:json];(rel["county:ansi"][boundary=administrative];);out tags;`

The interesting results should be in 'Data' tab.
Change `tags` to `geom`and you will have a shape data, matching exactly boundaries marked on openstreetmap.

The same queries work in real overpass, which should be used by software: `http://overpass-api.de/api/interpreter`

**Comment by kojoty** [comment]
@rapotek but we definitely need offline data: we need to be able to do for example: 
- search all caches in region "wojewodztwo pomorskie"

If we relay on external api and store the results for each geocache we can't be sure that we have consistent data (because external api data can be changed in time...) 

...,but if you know how to get the offline copy of OSM data with shapefiles - it would be useful for us as a source of geo-data for table.

**Comment by andrixnet** [comment]
I agree on the point that we need our own offline copy of the data, for an open-source, volunteer driven project to rely on third party on-line service providers is not feasible (in the long run).

@kojoty let's follow up with ZIP codes here: https://github.com/opencaching/opencaching-pl/issues/865

**Comment by andrixnet** [comment]
From which to actually use only the countries and subdivisions needed. 
500MB as OSM data? (might be different in SQL)

**Comment by rapotek** [comment]
@andrixnet a folder with SQLs ready to import: 556 MB, a mysql db folder after import: 434 MB. But it might be different for up-to-date data.

**Comment by kojoty** [comment]
if you know where i can download this data I would like to take a look closer on this - maybe we can just filter out some of the data from that - but I need to review it

**Comment by rapotek** [comment]
@kojoty Up until now I cannot locate original sql files with downloaded data (backuped it somewhere and forgot), but I will send you an email with link to gzipped mysqldump of above tables. Remember, codes and shapes were valid on 2018/11/11, where on 4835 iso3166-2 codes only 4206 shapes were available to download.

**Comment by kojoty** [comment]
> @kojoty Up until now I cannot locate original sql files with downloaded data (backuped it somewhere and forgot), but I will send you an email with link to gzipped mysqldump of above tables... 

cool - let me review this - thanks.
